pandas/core/indexes/datetimes.py

```diff
@@ -172,6 +172,50 @@ def _new_DatetimeIndex(cls, d):
     return result
 
 
+class Iterator:
+    def __init__(self, index):
+        self.data = index.asi8
```

Comment:
we already have a ``BaseIterator`` class in ``pandas.io.common``, it would need to be moved to ``pandas.core.common`` to use it here

Reply:
but why are you creating a class for this anyhow?

Reply:
Yes, and I'd prefer to keep this in `DatetimeIndex.__iter__` unless it's going to be used elsewhere.

Reply:
If `DatetimeIndex` class has `__next__` method and `__iter__` returns itself,  some other codes see `DatetimeIndex` as a iterator rather than a sequence. This cause some failures (#21012). So I wanted to hide `__next__` somehow. 

I found `BaseIterator`. But with it, it would become complicated because we have to make a subclass or override methods of a instance of `BaseIterator` (is this correct?).

Reply:
I think that most of the state tracking can be done inside `DatetimeIndex.__iter__`. Unless I'm misunderstanding, there's no need or desire to add `DatetimeIndex.__next__`.

Reply:
hmm... sorry, but I can't come up with the way using only `__iter__` and avoiding `yield` (#18848).